### PR TITLE
Implement logprob convenience for Liesel models

### PR DESCRIPTION
**Pull request by jobrachem**
This PR implements two classes that provide convenience for Liesel models: `lsl.LogProb` and `lsl.FlatLogProb`. The idea is to provide easy, direct access to a Liesel model's log prob, log lik and log prior in the form of functions that accept 

- a dictionary of variable names and corresponding values to use (`lsl.LogProb`)
- a flat array of variable values to use (`lsl.FlatLogProb`).

Why does this help? I like it a lot for debugging. These classes can also be used by users to build Liesel-specific inference tools. Fundamentally, the functionality is related to the `LieselInterface`. The main differences are: `lsl.LogProb` and `lsl.FlatLogProb` can also be used to single out the likelihood or the prior, and they accept position dictionaries instead of model states, which makes them more user-friendly. 

One could easily build something very similar in a general way for all `gs.ModelInterface` classes, i.e. something that accepts positions directly and offers gradients and hessians via Jax directly -- and maybe that is something we should do instead (or also). The `gs.ModelInterface` classes do, however, not distinguish between log prob, log lik and log prior, so they could only evaluate the log prob and its derivatives.

Here are examples:

```python
mu = lsl.Var.new_param(0.0, name="mu")
sigma = lsl.Var.new_param(1.0, name="sigma")
y = lsl.Var.new_obs(1.0, lsl.Dist(tfd.Normal, loc=mu, scale=sigma), name="y")
model = lsl.Model([y])
```

### LogProb

```python
logprob = lsl.LogProb(model, component="log_prob") 
# component could also be 'log_lik' or 'log_prior'

logprob({"mu": 1.0}) # Array(-0.9189385, dtype=float32)
logprob.log_prob({"mu": 1.0}) # Array(-0.9189385, dtype=float32) (same as direct call)
logprob.grad({"mu": 1.0}) # {'mu': Array(0., dtype=float32, weak_type=True)}
logprob.hessian({"mu": 1.0}) # {'mu': {'mu': Array(-1., dtype=float32, weak_type=True)}}
```

### FlatLogProb

```python
logprob = lsl.FlatLogProb(model, param_names=["mu"])
# component could also be 'log_lik' or 'log_prior'

logprob(jnp.array([1.0])) # Array(-0.9189385, dtype=float32)
logprob.log_prob(jnp.array([1.0])) # Array(-0.9189385, dtype=float32) (same as direct call)
logprob.grad(jnp.array([1.0])) # Array([0.], dtype=float32)
logprob.hessian(jnp.array([1.0])) # Array([[-1.]], dtype=float32)
```

I think providing both a flattened and a non-flattened version makes sense, because the first one (`lsl.LogProb`) may be easier to understand and use in many cases, while the second one `lsl.FlatLogProb` may be more handy to build upon in some cases.

### CODE REVIEW

File: tests/model/test_var.py [hunk 1]
```diff
@@ -766,3 +767,44 @@ def test_transform_default(self) -> None:
...
+        x = jax.random.uniform(jax.random.PRNGKey(1), (n,))
+        b = 1.0
+        e = jax.random.normal(jax.random.PRNGKey(2), (n,))
+        y = x * b + e
+
+        xvar = lnodes.Var.new_obs(x, name="x")
+        bvar = lnodes.Var.new_param(jnp.array([b]), name="b")
+        loc = lnodes.Var.new_calc(lambda x, b: x * b, x=xvar, b=bvar, name="loc")
+        scale = lnodes.Var.new_param(jnp.array([1.0]), name="scale")
+        scale.transform(tfp.bijectors.Exp())
```
[the remark this answers is not in the document]
Comment (jobrachem) [reply tests/model/test_var.py:784]: Spurious

File: tests/model/test_model.py [hunk 2]
```diff
@@ -351,6 +351,134 @@ def test_build_after_transform(self) -> None:
...
+        assert model.extract_position(["z"], model_state=state)["z"] == pytest.approx(
+            3.0
+        )
+        assert model.nodes["z"].value == pytest.approx(3.0)
+
+
+class TestPredictions:
+    def test_predict_at_current_state(self, model) -> None:
+        samples = {
+            "sigma_hat": tfd.Uniform().sample((4, 3), rnd.PRNGKey(6)),
```
[the remark this answers is not in the document]
Comment (jobrachem) [reply tests/model/test_model.py:390]: Spurious

**Comment by jobrachem** [comment]
Updated the PR. 

- Renamed `liesel.model.LogProb` -> `liesel.model.LieselLogProb` (same for the FlatLogProb class)
- Added `liesel.goose.LogProb`, which operates on any `gs.ModelInterface`. Simply gives convenient access to log probability, gradient, and hessian as functions based on a `Position`. There's also `liesel.goose.FlatLogProb`, which does the same but operates on position arrays and returns arrays.

This PR is now ready to be reviewed and looking for a reviewer.

Usecases are mainly easier debugging through easier access to gradients and hessians. But the `gs.LogProb` and `gs.FlatLogProb` classes can also be quite handy when implementing new estimation functionality.
